sleigh/sleigh-runtime/src/decoder.rs

```diff
@@ -204,9 +216,29 @@ impl Decoder {
                     let value = self.eval_context_expr(*expr, sleigh);
                     field.set(&mut self.context, value);
                 }
-                DecodeAction::SaveContext(field) => {
+                DecodeAction::SaveContext(field, expr) => {
                     let value = field.extract(self.context);
-                    field.set(&mut self.global_context, value);
+                    let addr = self.eval_context_expr(*expr, sleigh);
+                    tracing::trace!(
+                        "Saving context at 0x{:X} with value {}. PC is currently 0x{:X}.",
+                        addr,
+                        value,
+                        self.base_addr
+                    );
+
+                    // Modify global_context directly if change should take effect immediately.
+                    // Otherwise, store it for the future modification.
+                    if addr == self.base_addr as i64 {
+                        field.set(&mut self.global_context, value);
+                    } else if let Some(existing) = self
+                        .future_context_mods
+                        .iter_mut()
+                        .find(|(_, existing_field, _)| *existing_field == *field)
```

Comment:
See comment above, I'm not sure about this logic, this seems to force there to be only a single global set per field.